### PR TITLE
TCP Integration Testing

### PR DESCRIPTION
**Pull request by GeorgeHahn**
### What does this PR do?

Add baseline support for TCP integration tests

### Motivation

This will allow more of the lading functionality to be tested

### Related issues

Followup to #261

### CODE REVIEW

File: integration/ducks/src/main.rs [hunk 1]
```diff
@@ -113,107 +138,148 @@ async fn http_req_handler(req: Request<Body>) -> Result<hyper::Response<Body>, h
...
+            .timeout(Duration::from_secs(1))
+            .service(service);
+
+        let server = hyper::Server::builder(addr).serve(svc);
+        server.await?;
         Ok(())
     }
+
+    async fn tcp_handler(mut socket: TcpStream) -> Result<(), anyhow::Error> {
+        // 500KB input buffer per connection (this can probably be smaller)
```
Comment (blt) [inline integration/ducks/src/main.rs:251]: `KiB` properly, I think?

File: integration/ducks/src/main.rs [hunk 2]
```diff
@@ -113,107 +138,148 @@ async fn http_req_handler(req: Request<Body>) -> Result<hyper::Response<Body>, h
...
+            socket.read_buf(&mut buffer).await?;
+
+            {
+                let metric = TCP_COUNTERS.get().unwrap();
+                let mut m = metric.lock().await;
+                m.read_count += 1;
+                m.total_bytes += buffer.len() as u64;
+                m.entropy.add(entropy::metric_entropy(&buffer) as f64);
+
+                // TODO: data rate metrics & a way to measure consistency
```
Comment (blt) [inline integration/ducks/src/main.rs:267]: Confirming this was intentionally left. Seems like.

File: integration/sheepdog/src/lib.rs [hunk 3]
```diff
@@ -144,18 +162,32 @@ impl IntegrationTest {
...
+        let mut stdout_buf = String::new();
+        stdout.read_to_string(&mut stdout_buf).await?;
+        println!("ducks output:\n{}", stdout_buf);
+
+        let mut stdout = lading.stdout.unwrap();
+        let mut stdout_buf = String::new();
+        stdout.read_to_string(&mut stdout_buf).await?;
+        println!("lading output:\n{}", stdout_buf);
+
+        println!("test result: (todo)");
```
Comment (blt) [inline integration/sheepdog/src/lib.rs:186]: Might be worth pulling `tracing` in at some point to handle output?

**Review by blt — APPROVED** [review]
Very nice. I had some comments, a question. I think when you're ready a bit in the README about how to add integration tests to the project is warranted.

File: integration/sheepdog/src/lib.rs (hunk 3 above)
Comment (GeorgeHahn) [reply integration/sheepdog/src/lib.rs:186]: I don't think we need `tracing` here yet, but we will before too long. These logs are emitted by `tracing` in the child, so they're already formatted nicely. This print gets them into cargo's test harness where they're only visible if the `--no-capture` flag is passed in.

This flow is going to change when we switch to RPC-streamed logs to support lading's binary-launch mode.

File: integration/ducks/src/main.rs (hunk 2 above)
Comment (GeorgeHahn) [reply integration/ducks/src/main.rs]: It was, but I removed it after thinking about it some more. I need to read up on data rate calculations and may have some questions in our team channel. This isn't a terribly high priority at the moment and we might be able to get by with chunk size unit tests in lading.

**Comment by GeorgeHahn** [comment]
I haven't been able to get a local repro for this failure; this even runs successfully in [act](https://github.com/nektos/act).

In Actions:
- Lading & ducks both start correctly
- Both run for the expected duration

I'm going to need to debug in-branch unfortunately. Sorry in advance for the noise.

**Comment by blt** [comment]
> I'm going to need to debug in-branch unfortunately. Sorry in advance for the noise.

All good! Commits are free and I can always unsubscribe if it gets real noisy.

**Comment by GeorgeHahn** [comment]
`lading` was trying to connect to ipv6 localhost 😭

**Review by blt — APPROVED** [review]

**Pull request merged**